### ISSUE TITLE
Tracking opened connections by ExoPlayer/OkHttp. Leaks.

### ISSUE THREAD

**Issue by stari4ek** [google/ExoPlayer#7837]
### [REQUIRED] Searched documentation and issues
n/a

### [REQUIRED] Question
My app plays (mostly) live content (hls/ss/mpd/http progressive) using ExoPlayer.
When user switches to next media stream - new `MediaSource` is created and player is prepared with it.
I'm trying to limit number of opened overlapped connections when switching from one media stream to another one.
Next stream is unknown before the moment when user requests switching.

To achieve this requirement I'm tracking opened connections using okhttp `EventListener.connectionAcquired`/`EventListener.connectionReleased`. 

Steps:
1. user requests playback of stream1
1. App creates `MediaSource` + `SimpleExoPlayer.prepare()`
1. user watches for N sec
1. user requests to play stream2
1. `SimpleExoPlayer.stop(reset=true)` or `SimpleExoPlayer.clearVideoSurface()` + `SimpleExoPlayer.release()` is called
1. App is waiting when there is no opened connections (timeout: 3 sec)
1. App creates `MediaSource` + `SimpleExoPlayer.prepare()`

I have tried lightweight "stop" and full "release" of player, but in both cases I'm seeing that there are a lot of 3 sec timeout happening, while waiting for all opened connections to be released.

I cannot reproduce it locally with devices and streams I have, but there are pretty big percent of users facing this issue daily. I do not see any correlation between devices, android versions or stream types.

Does it look like connections leaks in ExoPlayer or I misuse _okhttp_/_ExoPlayer_ itself? 
Maybe 3 sec is too small to fully stop loading and release connections (I had 10 sec timeout originally. same issues.)

My current ExoPlayer version is based on: 81b0b53a37d437a8c5933208070e92a5942dbc03

### A full bug report captured from the device
n/a

### Link to test content
n/a

**Comment by ojw28** [comment]
> I'm trying to limit number of opened overlapped connections when switching from one media stream to another one.

Why? It sounds a little unnecessary, and to be causing fairly significant problems that you could avoid simply by not trying to do this at all.

> Does it look like connections leaks in ExoPlayer or I misuse okhttp/ExoPlayer itself?

ExoPlayer will close the response body for any requests that are in flight (see `OkHttpDataSource` for the implementation), which I think is what we're supposed to do. This is done asynchronously, but should complete quickly under normal circumstances.

I don't know what OkHttp does underneath us. It could be that OkHttp isn't very proactive about closing the connection, either on purpose (e.g., I would hypothesise that OkHttp implements socket re-use) or due to design limitations. It's unclear to me how socket re-use affects the semantics of `connectionAcquired` and `connectionReleased`.

To summarise:

* I think we're doing the right thing at the ExoPlayer layer. If you think we should be doing something else, please let us know!
* It's not completely clear what's going on in the OkHttp layer, or whether it's working as intended or not.
* I would suggest you consider simply not doing what you're trying to do, as a way of solving this problem.

**Comment by stari4ek** [comment]
Thank you for your prompt response.

> Why? It sounds a little unnecessary, and to be causing fairly significant problems that you could avoid simply by not trying to do this at all.

There are IPTV services which apply connections limits/quota. It's hard to tell how exactly they count connections (opened sockets? requests rate?..), but without this "waiting" there is much higher chance to get `403 Forbidden` from such services when switching from one stream to another.

**Comment by ojw28** [comment]
Thanks for the information. IPTV services applying connection limits sounds like something that's outside of our control. As an aside, it also doesn't sound particularly sensible, unless it's deliberately designed to stop you from doing what you're trying to do.

As above, I think we're doing the right thing at the ExoPlayer layer, so I think we can close this.

**Comment by stari4ek** [comment]
It looks like I found the problem, when http connection lives longer than it's ExoPlayer instance.
I've reproduced it locally, when there is timeout for connection and `LoadErrorHandlingPolicy` grants retry.

In my case:

1. preparing with stream1
2. connection timeout. retries.
3. while retrying ExoPlayer instance is **released**.
4. Start playback with new Exoplayer instance for different stream.
5. Get `OkHttpDataSource.open` for connection from already released ExoPlayer

Log:
```
2020-09-18 14:55:02.252 12968-12968/by.stari4ek.tvirl.debug D/MediaSourceFactory: Building media source for
        MediaSourceInfo{streamType=Hls, src=https://devstreaming-cdn.apple.com/videos/streaming/examples/bipbop_4x3/bipbop_4x3_variant.m3u8, drm=DrmInfo{drmUuid=edef8ba9-79d6-4ace-a3c8-27dcd51d21ed, licenseUri=https://proxy.uat.widevine.com/proxy?video_id=GTS_SW_SECURE_CRYPTO&provider=widevine_test, licenseRequestHeaders=}}

2020-09-18 14:55:14.462 12968-13170/by.stari4ek.tvirl.debug D/TvirlLoadErrorHandlingPolicy: Got HttpDataSourceException (Unable to connect) (count: 1, load duration: 10108ms). Delay reconnect to: 0ms
2020-09-18 14:55:14.465 12968-12968/by.stari4ek.tvirl.debug E/EventLogger: internalError [eventTime=12.23, mediaPos=0.00, window=0, loadError
      com.google.android.exoplayer2.upstream.HttpDataSource$HttpDataSourceException: Unable to connect
        at com.google.android.exoplayer2.ext.okhttp.OkHttpDataSource.open(OkHttpDataSource.java:225)
        at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:177)
        at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:83)
        at com.google.android.exoplayer2.upstream.DataSourceInputStream.checkOpened(DataSourceInputStream.java:101)
        at com.google.android.exoplayer2.upstream.DataSourceInputStream.open(DataSourceInputStream.java:64)
        at com.google.android.exoplayer2.upstream.ParsingLoadable.load(ParsingLoadable.java:177)
        at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:413)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.net.SocketTimeoutException: timeout
        at okio.SocketAsyncTimeout.newTimeoutException(Okio.kt:159)
        at okio.AsyncTimeout.exit$jvm(AsyncTimeout.kt:203)
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:163)
        at okio.RealBufferedSource.indexOf(RealBufferedSource.kt:349)
        at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.kt:222)
        at okhttp3.internal.http1.Http1ExchangeCodec.readHeaderLine(Http1ExchangeCodec.java:242)
        at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:213)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at by.stari4ek.okhttp.BasicAuthInterceptor.intercept(BasicAuthInterceptor.java:43)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at by.stari4ek.okhttp.UserAgentInterceptor.intercept(UserAgentInterceptor.java:38)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at by.stari4ek.iptv4atv.player.source.MasterHttpDataSourceTracker.intercept(MasterHttpDataSourceTracker.java:38)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at by.stari4ek.iptv4atv.player.source.StreamTypeValidatingInterceptor.intercept(StreamTypeValidatingInterceptor.java:59)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
          at okhttp3.internal.http.RealInterceptorChain


2020-09-18 14:55:15.571 12968-12968/by.stari4ek.tvirl.debug I/ExoPlayerImpl: Release a0612a6 [ExoPlayerLib/2.11.6] [generic_x86, sdk_google_atv_x86, unknown, 27] [goog.exo.core, goog.exo.hls, goog.exo.okhttp, goog.exo.ffmpeg]


2020-09-18 14:55:18.583 12968-12968/by.stari4ek.tvirl.debug I/ExoPlayerImpl: Init 426ed57 [ExoPlayerLib/2.11.6] [generic_x86, sdk_google_atv_x86, unknown, 27]

2020-09-18 14:55:18.592 12968-12968/by.stari4ek.tvirl.debug D/MediaSourceFactory: Building media source for
        MediaSourceInfo{streamType=Hls, src=<REDACTED>, }


CALLSTACK from breakpoint:

java.lang.Exception: Stack trace
    at java.lang.Thread.dumpStack(Thread.java:1348)
    at by.stari4ek.iptv4atv.player.source.StreamTypeValidatingInterceptor.validate(StreamTypeValidatingInterceptor.java:81)
    at by.stari4ek.iptv4atv.player.source.StreamTypeValidatingInterceptor.intercept(StreamTypeValidatingInterceptor.java:65)
    at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
    at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
    at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
    at okhttp3.RealCall.execute(RealCall.java:81)
    at com.google.android.exoplayer2.ext.okhttp.OkHttpDataSource.open(OkHttpDataSource.java:220)
    at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:177)
    at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:83)
    at com.google.android.exoplayer2.upstream.DataSourceInputStream.checkOpened(DataSourceInputStream.java:101)
    at com.google.android.exoplayer2.upstream.DataSourceInputStream.open(DataSourceInputStream.java:64)
    at com.google.android.exoplayer2.upstream.ParsingLoadable.load(ParsingLoadable.java:177)
    at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:413)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
    at java.lang.Thread.run(Thread.java:764)
```

At this moment there are 2 `DefaultHlsPlaylistTracker:MediaPlaylist` threads.

According to  the context of `Loader.run` in this stack:
```
shouldLoad = true;
canceled = true
```

So it looks like there was blocking I/O (open), and while it was waiting - Loader become cancelled.

**Comment by ojw28** [comment]
I'm not really sure what logic you're using to confirm that the stack trace at the bottom of your post above is coming from the previous player instance.

It looks like there can be multiple `DefaultHlsPlaylistTracker:MediaPlaylist` threads owned by a single player, so I don't think the presence of two of them tells you anything about which player owns them. I did take a look at the code around what happens when the player is released, and it looks like everything is released correctly, including the case where a request is pending retry.

So, I don't think there's evidence here that's persuasive enough to warrant reopening the issue. If you think there's a genuine issue here, you should be able to reproduce it using only a single player. Specifically, you should be able to release a single player and see an `OkHttpDataSource.open` call after it's been released. I'd suggest you see whether it's possible to reproduce this way, and ideally whether it's possible to come up with reliable reproduction steps (e.g., by making a code change to simulate the failure/retry in a reproducible way). That would provide a strong enough case for investigating this further.

**Comment by stari4ek** [comment]
I'm 100% sure that callstack was from the "previous" ExoPlayer instance (technically, it's not exoplayer, but Loader thread it spawned). It's `DataSpec` with uri belonged to first media stream.
I do not think that ExoPlayer has issue here, and it's behavior looks valid. ExoPlayer cannot cancel  blocking `open` immediately, so thread is spinning while "old" ExoPlayer is released and new one created and prepared. Those spinning thread will be stopped when blocking `open` is finished.
`LoadTask.cancel` calls `Thread.interrupt`, but most probably there is no `wait`/`sleep`, so nothing to interrupt.

I will try to investigate it further. No need to open this issue (at least yet). Just trying to unwrap the story.
